rest/data/project_settings.go

```diff
@@ -97,6 +97,97 @@ func disableStartingSettings(p *model.ProjectRef) {
 	p.CommitQueue.Enabled = utility.FalsePtr()
 }
 
+// PromoteVarsToRepo moves variables from an attached project to its repo.
+// Promoted vars are removed from the project as part of this operation.
+// Variables whose names already appear in the repo settings will be overwritten.
+func PromoteVarsToRepo(projectId string, varNames []string, userId string) error {
+	project, err := model.GetProjectSettingsById(projectId, false)
+	if err != nil {
+		return errors.Wrapf(err, "getting project settings for project '%s'", projectId)
+	}
+	projectVars, err := model.FindOneProjectVars(projectId)
+	if err != nil {
+		return errors.Wrapf(err, "getting project variables for project '%s'", projectId)
+	}
+
+	repo, err := model.GetProjectSettingsById(project.ProjectRef.RepoRefId, true)
+	if err != nil {
+		return errors.Wrapf(err, "getting repo settings for repo '%s'", project.ProjectRef.RepoRefId)
+	}
+	repoVars, err := model.FindOneProjectVars(project.ProjectRef.RepoRefId)
+	if err != nil {
+		return errors.Wrapf(err, "getting repo variables for repo '%s'", project.ProjectRef.RepoRefId)
+	}
+
+	// Add each promoted variable to existing repo vars
+	apiRepoVars := &restModel.APIProjectVars{}
+	apiRepoVars.BuildFromService(*repoVars)
+	for _, varName := range varNames {
+		// Ignore nonexistent variables
+		if _, contains := projectVars.Vars[varName]; !contains {
```

Comment:
nit: usually, the second variable from a map lookup just called `ok` (the ["comma-ok" idiom](https://go.dev/doc/effective_go#maps)) just like how error types are `err` unless you have a special need for a different variable name.